### DISCUSSION TITLE
How to implement a state-variable-dependent and stepwise coefficient in PDE

### DISCUSSION THREAD

**Discussion by XyM6u9** [lululxvi/deepxde#1416] [Q&A] [answered]
Such as Richards equation:

`dHdx =  dde.grad.jacobian(y, x, i=0, j=0)`
`d2kx = dde.grad.jacobian(K*dHdx, x, i=0, j=0)`

`dHdy =  dde.grad.jacobian(y, x, i=0, j=1)`
`d2ky = dde.grad.jacobian(K*dHdy, x, i=0, j=1)`
      
`return d2kx + d2ky`
if H<0
K= ks*kr(H) 
else 
K = ks

my code:

```
def pde(x, y):    
    uw = (y-x[:, 1:2])*1000*9.8
    kr = (1 - tf.abs(-k_a * uw) ** (sr_n * sr_m) / ((1 + tf.abs(-k_a * uw) ** sr_n) ** sr_m)) ** 2 / (1 + tf.abs(-k_a * uw) ** sr_n) ** (sr_m / 2)
    
    K = tf.cond(uw<0, lambda: tf.multiply(ks, kr), lambda: ks)         

    dhdx =  dde.grad.jacobian(y, x, i=0, j=0)
    d2kx = dde.grad.jacobian(K*dhdx, x, i=0, j=0)
    
    dhdy =  dde.grad.jacobian(y, x, i=0, j=1)
    d2ky = dde.grad.jacobian(K*dhdy, x, i=0, j=1)
          
    return d2kx + d2ky
```
 Traceback (most recent call last):

  File ~\anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:88
    model.compile("adam", lr=0.002)

  File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:22 in wrapper
    result = f(*args, **kwargs)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:132 in compile
    self._compile_tensorflow_compat_v1(lr, loss_fn, decay, loss_weights)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:181 in _compile_tensorflow_compat_v1
    losses_train = losses(self.data.losses_train)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:167 in losses
    losses = losses_fn(

  File ~\anaconda3\lib\site-packages\deepxde\data\data.py:13 in losses_train
    return self.losses(targets, outputs, loss_fn, inputs, model, aux=aux)

  File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:140 in losses
    f = self.pde(inputs, outputs_pde)

  File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:23 in pde
    K = tf.cond(uw<0, lambda: tf.multiply(ks, kr), lambda: ks)

  File ~\anaconda3\lib\site-packages\tensorflow\python\util\traceback_utils.py:153 in error_handler
    raise e.with_traceback(filtered_tb) from None

  File ~\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py:1751 in _create_c_op
    raise ValueError(e.message)

ValueError: Shape must be rank 0 but is rank 2 for '{{node cond_1/Switch}} = Switch[T=DT_BOOL](Less_2, Less_2)' with input shapes: [?,1], [?,1].   
    
-----------------------------------------------------------------------------------------------------------------------------
try to use `auxiliary_var_function` still doesnt work

```
uw = np.linspace(-100, 200, 300)
K = (ks * (1 - np.abs(-k_a * uw) ** (sr_n * sr_m) / ((1 + np.abs(-k_a * uw) ** sr_n) ** sr_m)) ** 2 / ((1 + np.abs(-k_a * uw) ** sr_n) ** (sr_m / 2))) * (uw < 0) + ks * (uw >= 0)

def ex_func(pwp):
    spline = sp.interpolate.Rbf(
        uw, K, function="thin_plate", smooth=0, episilon=0
    )
    return spline(pwp)


def pde(x, y):    
    pwp = (y-x[:, 1:])*1*9.8 #kPa            

    dHdx =  dde.grad.jacobian(y, x, i=0, j=0)
    d2kx = dde.grad.jacobian(ex_func(pwp)*dHdx, x, i=0, j=0)
    
    dHdy =  dde.grad.jacobian(y, x, i=0, j=1)
    d2ky = dde.grad.jacobian(ex_func(pwp)*dHdy, x, i=0, j=1)
          
    return d2kx + d2ky 

```

  File ~\anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:98
    model.compile("adam", lr=0.002)

  File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:22 in wrapper
    result = f(*args, **kwargs)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:132 in compile
    self._compile_tensorflow_compat_v1(lr, loss_fn, decay, loss_weights)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:181 in _compile_tensorflow_compat_v1
    losses_train = losses(self.data.losses_train)

  File ~\anaconda3\lib\site-packages\deepxde\model.py:167 in losses
    losses = losses_fn(

  File ~\anaconda3\lib\site-packages\deepxde\data\data.py:13 in losses_train
    return self.losses(targets, outputs, loss_fn, inputs, model, aux=aux)

  File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:140 in losses
    f = self.pde(inputs, outputs_pde)

  File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:41 in pde
    d2kx = dde.grad.jacobian(ex_func(pwp)*dHdx, x, i=0, j=0)

  File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:34 in ex_func
    return spline(pwp)

  File ~\anaconda3\lib\site-packages\scipy\interpolate\_rbf.py:280 in __call__
    args = [np.asarray(x) for x in args]

  File ~\anaconda3\lib\site-packages\scipy\interpolate\_rbf.py:280 in <listcomp>
    args = [np.asarray(x) for x in args]

  File ~\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py:836 in __array__
    raise NotImplementedError(

NotImplementedError: Cannot convert a symbolic tf.Tensor (mul_5:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported.

**Comment by lululxvi** [comment, marked as answer]
Just implement the function using TensorFlow/PyTorch.

**Comment by BraveDrXuTF** [comment]
> Such as Richards equation:
> 
> `dHdx = dde.grad.jacobian(y, x, i=0, j=0)` `d2kx = dde.grad.jacobian(K*dHdx, x, i=0, j=0)`
> 
> `dHdy = dde.grad.jacobian(y, x, i=0, j=1)` `d2ky = dde.grad.jacobian(K*dHdy, x, i=0, j=1)`
> 
> `return d2kx + d2ky` if H<0 K= ks*kr(H) else K = ks
> 
> my code:
> 
> ```
> def pde(x, y):    
>     uw = (y-x[:, 1:2])*1000*9.8
>     kr = (1 - tf.abs(-k_a * uw) ** (sr_n * sr_m) / ((1 + tf.abs(-k_a * uw) ** sr_n) ** sr_m)) ** 2 / (1 + tf.abs(-k_a * uw) ** sr_n) ** (sr_m / 2)
>     
>     K = tf.cond(uw<0, lambda: tf.multiply(ks, kr), lambda: ks)         
> 
>     dhdx =  dde.grad.jacobian(y, x, i=0, j=0)
>     d2kx = dde.grad.jacobian(K*dhdx, x, i=0, j=0)
>     
>     dhdy =  dde.grad.jacobian(y, x, i=0, j=1)
>     d2ky = dde.grad.jacobian(K*dhdy, x, i=0, j=1)
>           
>     return d2kx + d2ky
> ```
> 
> Traceback (most recent call last):
> 
> File ~\anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec exec(code, globals, locals)
> 
> File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:88 model.compile("adam", lr=0.002)
> 
> File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:22 in wrapper result = f(*args, **kwargs)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:132 in compile self._compile_tensorflow_compat_v1(lr, loss_fn, decay, loss_weights)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:181 in _compile_tensorflow_compat_v1 losses_train = losses(self.data.losses_train)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:167 in losses losses = losses_fn(
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\data.py:13 in losses_train return self.losses(targets, outputs, loss_fn, inputs, model, aux=aux)
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:140 in losses f = self.pde(inputs, outputs_pde)
> 
> File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:23 in pde K = tf.cond(uw<0, lambda: tf.multiply(ks, kr), lambda: ks)
> 
> File ~\anaconda3\lib\site-packages\tensorflow\python\util\traceback_utils.py:153 in error_handler raise e.with_traceback(filtered_tb) from None
> 
> File ~\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py:1751 in _create_c_op raise ValueError(e.message)
> 
> ValueError: Shape must be rank 0 but is rank 2 for '{{node cond_1/Switch}} = Switch[T=DT_BOOL](Less_2, Less_2)' with input shapes: [?,1], [?,1].
> 
> try to use `auxiliary_var_function` still doesnt work
> 
> ```
> uw = np.linspace(-100, 200, 300)
> K = (ks * (1 - np.abs(-k_a * uw) ** (sr_n * sr_m) / ((1 + np.abs(-k_a * uw) ** sr_n) ** sr_m)) ** 2 / ((1 + np.abs(-k_a * uw) ** sr_n) ** (sr_m / 2))) * (uw < 0) + ks * (uw >= 0)
> 
> def ex_func(pwp):
>     spline = sp.interpolate.Rbf(
>         uw, K, function="thin_plate", smooth=0, episilon=0
>     )
>     return spline(pwp)
> 
> 
> def pde(x, y):    
>     pwp = (y-x[:, 1:])*1*9.8 #kPa            
> 
>     dHdx =  dde.grad.jacobian(y, x, i=0, j=0)
>     d2kx = dde.grad.jacobian(ex_func(pwp)*dHdx, x, i=0, j=0)
>     
>     dHdy =  dde.grad.jacobian(y, x, i=0, j=1)
>     d2ky = dde.grad.jacobian(ex_func(pwp)*dHdy, x, i=0, j=1)
>           
>     return d2kx + d2ky 
> ```
> 
> File ~\anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec exec(code, globals, locals)
> 
> File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:98 model.compile("adam", lr=0.002)
> 
> File ~\anaconda3\lib\site-packages\deepxde\utils\internal.py:22 in wrapper result = f(*args, **kwargs)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:132 in compile self._compile_tensorflow_compat_v1(lr, loss_fn, decay, loss_weights)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:181 in _compile_tensorflow_compat_v1 losses_train = losses(self.data.losses_train)
> 
> File ~\anaconda3\lib\site-packages\deepxde\model.py:167 in losses losses = losses_fn(
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\data.py:13 in losses_train return self.losses(targets, outputs, loss_fn, inputs, model, aux=aux)
> 
> File ~\anaconda3\lib\site-packages\deepxde\data\pde.py:140 in losses f = self.pde(inputs, outputs_pde)
> 
> File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:41 in pde d2kx = dde.grad.jacobian(ex_func(pwp)*dHdx, x, i=0, j=0)
> 
> File c:\users\37601\onedrive\桌面\pinn\unsat2d.py:34 in ex_func return spline(pwp)
> 
> File ~\anaconda3\lib\site-packages\scipy\interpolate_rbf.py:280 in **call** args = [np.asarray(x) for x in args]
> 
> File ~\anaconda3\lib\site-packages\scipy\interpolate_rbf.py:280 in args = [np.asarray(x) for x in args]
> 
> File ~\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py:836 in **array** raise NotImplementedError(
> 
> NotImplementedError: Cannot convert a symbolic tf.Tensor (mul_5:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported.

So how has it exactly been solved? Could you offer your code snippet with success?

**Reply by notviral** [reply]
did you got any success on solving this ?
